### PR TITLE
Rn 588 data broker tei

### PR DESCRIPTION
**Pull request by IgorNadj**
RN-588

A bunch of refactors around testing, and then two last commits contain:
- db change to allow finding data with json paths
- data-broker change which adds entity codes when the DHIS response has a "tei" column

### CODE REVIEW

File: packages/database/src/testUtilities/createModelsStub.js [hunk 1]
```diff
@@ -32,11 +32,12 @@ const mockFind = (array, criteria = {}) =>
 
       if (key.includes('->>')) {
         // Special case
-        const lookupPath = key.split('->>');
-        if (lookupPath.length > 2) throw new Error('Not implemented yet...');
```
Comment (edmofro) [inline packages/database/src/testUtilities/createModelsStub.js]: Hooray, implemented now!

File: packages/data-broker/src/services/dhis/builders/addTrackedEntityCodes.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export const addTrackedEntityCodes = async (models, events) => {
+  return events;
+};
```
Comment (edmofro) [inline packages/data-broker/src/services/dhis/builders/addTrackedEntityCodes.js:3]: Unused?

File: packages/data-broker/src/services/dhis/translators/TrackedEntityInstanceTranslator.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+/**
+ * Data pulls sometimes return tracked entity instance ids. This translator
+ * converts these ids to entity codes.
+ */
+export class TrackedEntityInstanceTranslator {
```
Comment (edmofro) [inline packages/data-broker/src/services/dhis/translators/TrackedEntityInstanceTranslator.js:10]: Unused?

**Review by edmofro — CHANGES_REQUESTED** [review]
Early review on the two meaningful commits - looking good so far, will review the PR as a whole now

File: packages/data-broker/src/services/dhis/builders/addTrackedEntityCodes.js (hunk 2 above)
Comment (IgorNadj) [reply packages/data-broker/src/services/dhis/builders/addTrackedEntityCodes.js]: Ah didn't spot this in my once-over, removed

File: packages/data-broker/src/services/dhis/translators/TrackedEntityInstanceTranslator.js (hunk 3 above)
Comment (IgorNadj) [reply packages/data-broker/src/services/dhis/translators/TrackedEntityInstanceTranslator.js]: Removed

File: packages/data-broker/src/services/dhis/translators/DhisCodeToIdTranslator.js [hunk 4]
```diff
@@ -10,7 +10,14 @@ import {
...
 
-export class DhisInputSchemeResolvingApiProxy {
+/**
+ * DhisCodeToIdTranslator can be used for data pulls from a DHIS instance where there are no codes set against
+ * the Data Elements, Org Units etc.
+ *
+ * It detects such instances, and swaps all the codes it is given to ids for the fetch, then does the same for
+ * the data coming back (swapping ids in the response data for codes).
+ */
+export class DhisCodeToIdTranslator {
```
Comment (edmofro) [inline packages/data-broker/src/services/dhis/translators/DhisCodeToIdTranslator.js:20]: Huge name improvement, and great comment above

File: packages/database/src/tests/testUtilities/createModelsStub.test.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { expect } from 'chai';
+import { createModelsStub } from '../../testUtilities';
+
+describe('createModelsStub', () => {
```
Comment (edmofro) [inline packages/database/src/tests/testUtilities/createModelsStub.test.js:9]: So meta

File: packages/database/src/testUtilities/createModelsStub.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    return getNestedProp(rest, obj[first]);
+  }
+  return obj[first] ?? undefined;
+};
+
+/**
+ * @param {MockData} mockData
+ * @return {} a partial implementation of ModelRegistry, to be used in its place
+ */
+export const createModelsStub = mockData => {
```
Comment (edmofro) [inline packages/database/src/testUtilities/createModelsStub.js:63]: Really good idea doing this

**Review by edmofro — APPROVED** [review]
There's some great deep work in here, thanks for doing so much code quality improvement as part of a hotfix

**Comment by IgorNadj** [comment]
Rebased on master so we can deploy this separately as a hotfix, the only commits that were dropped out of this PR were:
* https://github.com/beyondessential/tupaia/commit/f792752dec8058b20706598b204f2ecd8b723337
* https://github.com/beyondessential/tupaia/commit/35a53cc188d354851b1e390748224a05168abe9b
* https://github.com/beyondessential/tupaia/commit/676f375359a0c5cbc66a518b9f5f95cbba841327
* https://github.com/beyondessential/tupaia/commit/4c0466f0bce22bcef62981323ea20e550230fce6

These were part of the feature branch when testing. Looking through them, none of them have any changes related to DHIS or data pulling, so should be safe to pull out.

**Pull request merged**